nav2_map_server/src/map_saver.cpp

```diff
@@ -30,221 +31,173 @@
+      save_parameters.occupied_thresh = occupied_thresh_default_;
     }
 
-    RCLCPP_INFO(get_logger(), "Waiting for the map");
-    map_sub_ = create_subscription<nav_msgs::msg::OccupancyGrid>(
-      "map", rclcpp::SystemDefaultsQoS(),
+    // Add new subscription for incoming map topic
+    auto map_sub = map_listener_->create_subscription<nav_msgs::msg::OccupancyGrid>(
+      map_topic, rclcpp::SystemDefaultsQoS(),
       std::bind(&MapSaver::mapCallback, this, std::placeholders::_1));
-  }
-}
 
-void MapSaver::try_write_map_to_file(const nav_msgs::msg::OccupancyGrid & map)
-{
-  auto logger = get_logger();
-  RCLCPP_INFO(
-    logger, "Received a %d X %d map @ %.3f m/pix", map.info.width, map.info.height,
-    map.info.resolution);
-
-  std::string mapdatafile = mapname_ + "." + image_format;
-  {
-    // should never see this color, so the initialization value is just for debugging
-    Magick::Image image({map.info.width, map.info.height}, "red");
-
-    // In scale mode, we need the alpha (matte) channel. Else, we don't.
-    // NOTE: GraphicsMagick seems to have trouble loading the alpha channel when saved with
-    // Magick::GreyscaleMatte, so we use TrueColorMatte instead.
-    image.type(map_mode == MapMode::Scale ? Magick::TrueColorMatteType : Magick::GrayscaleType);
-
-    // Since we only need to support 100 different pixel levels, 8 bits is fine
-    image.depth(8);
-
-    for (size_t y = 0; y < map.info.height; y++) {
-      for (size_t x = 0; x < map.info.width; x++) {
-        int8_t map_cell = map.data[map.info.width * (map.info.height - y - 1) + x];
-
-        Magick::Color pixel;
-
-        switch (map_mode) {
-          case MapMode::Trinary:
-            if (map_cell < 0 || 100 < map_cell) {
-              pixel = Magick::ColorGray(205 / 255.0);
-            } else if (map_cell <= threshold_free_) {
-              pixel = Magick::ColorGray(254 / 255.0);
-            } else if (threshold_occupied_ <= map_cell) {
-              pixel = Magick::ColorGray(0 / 255.0);
-            } else {
-              pixel = Magick::ColorGray(205 / 255.0);
-            }
-            break;
-          case MapMode::Scale:
-            if (map_cell < 0 || 100 < map_cell) {
-              pixel = Magick::ColorGray{0.5};
-              pixel.alphaQuantum(TransparentOpacity);
-            } else {
-              pixel = Magick::ColorGray{(100.0 - map_cell) / 100.0};
-            }
-            break;
-          case MapMode::Raw:
-            Magick::Quantum q;
-            if (map_cell < 0 || 100 < map_cell) {
-              q = MaxRGB;
-            } else {
-              q = map_cell / 255.0 * MaxRGB;
-            }
-            pixel = Magick::Color(q, q, q);
-            break;
-          default:
-            throw std::runtime_error("Invalid map mode");
-        }
-        image.pixelColor(x, y, pixel);
+    rclcpp::Time start_time = now();
+    while (rclcpp::ok()) {
+      rclcpp::spin_some(map_listener_);
```

Comment:
If you use the `rclcpp_node_` above, you don't need to spin here

Reply:
Fixed: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-99885e6562ca50edd8cc80db6362d317L175